keps/sig-api-machinery/1027-api-unions/README.md

```diff
@@ -86,42 +120,133 @@ become possible.
 
 ### Goals
 
-The goal is to enable a union or "oneof" semantics in Kubernetes types, both for
-in-tree types and for CRDs.
+Provide a simple mechanism for API authors to label fields of a resource as
+members of a oneOf, in order to receive standardized:
+
+* Validation - ensuring only one member field is set (or at most one if
+  desired).
+* Normalization - ensuring the API server can modify the fields of the oneOf to
```

Comment:
Is this the goal? Or are we merely trying to detect and fail requests sent by clients who have misunderstood?

Reply:
There's no way for unstructured clients to get the object right without clearing the fields with version skew though.

Reply:
and the question is, do we force users to update such clients by failing their requests, or do we want to try and repair the requests.

Reply:
Right, in the former case, it means that adding a branch to a union is a backwards incompatible change, which I think is a little unfortunate (how do we even update Kubernetes if we start failing clients? We basically can't, so adding a branch to a union will be forbidden in practice).

Reply:
> Is this the goal? Or are we merely trying to detect and fail requests sent by clients who have misunderstood?

I'm also unsure that this is a goal versus detection and good failure messages.

> There's no way for unstructured clients to get the object right without clearing the fields with version skew though.

An unstructured client has access to all the fields in the union, right?  Is it a question of knowledge?


I think an explanation that include if and how server-side-apply should handle unions may help sway me one way or the other.  I'll keep reading below to see how that is addressed.

Reply:
I narrowed this description down to just talking about normalizing for clients unaware of fields due to version skew.